File: jointModel.py

# This code is me attempting to fix the issues that jointModel.R had with the C-index calculation
# The original code used the survcomp package, which is no longer maintained and has compatibility issues with recent R versions.
# Additionally, the analysis was slow and was not able to optimize for certain scoring methods.
# Although python does not have a direct equivalent to the JMbayes2 package in R, there is a system where
# I can create my own joint models using libraries like lifelines for survival analysis and statsmodels for longitudinal data.

# Load necessary packages
import pandas as pd
import numpy as np
from lifelines import CoxPHFitter
import pymc as pm
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.model_selection import KFold, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.metrics import make_scorer

# Load the dataset
data = pd.read_csv("onset_longitudinal.csv")

# Preprocess the data
data = data.dropna()
data['NR_Name'] = data['NR_Name'].astype('category')
data['sex'] = data['sex'].astype('category')
data['diet'] = data['diet'].astype('category')
data['Event'] = data['Event'].astype(int)
data['Event_time'] = data['Event_time'].astype(float)
data['week'] = data['week'].astype(float)
data['weight'] = data['weight'].astype(float)
data['rbg'] = data['rbg'].astype(float)

# Define custm joint model function
def joint_model(train_long, train_surv, test_long, test_surv):
    # Create a mixed effects model for the longitudinal data
    with pm.Model() as joint_model:
        # Priors for fixed effects
        beta_weight = pm.Normal('beta_weight', mu=0, sigma=10)
        beta_week = pm.Normal('beta_week', mu=0, sigma=10)
        beta_sex = pm.Normal('beta_sex', mu=0, sigma=10)
        beta_diet = pm.Normal('beta_diet', mu=0, sigma=10)
        sigma = pm.HalfNormal('sigma', sigma=1)

        # Random effects
        unique_ids = train_long['NR_Name'].cat.categories
        n_ids = len(unique_ids)
        u_intercept = pm.Normal('u_intercept', mu=0, sigma=1, shape=n_ids)
        u_weight = pm.Normal('u_weight', mu=0, sigma=1, shape=n_ids)
        u_week = pm.Normal('u_week', mu=0, sigma=1, shape=n_ids)
        u_sex = pm.Normal('u_sex', mu=0, sigma=1, shape=n_ids)
        u_diet = pm.Normal('u_diet', mu=0, sigma=1, shape=n_ids)
        id_idx = train_long['NR_Name'].cat.codes.values

        # Expected value
        mu = (beta_weight * train_long['weight'].values +
              beta_week * train_long['week'].values +
              beta_sex * train_long['sex'].values +
              beta_diet * train_long['diet'].values +
              u_intercept[id_idx] +
              u_weight[id_idx] * train_long['weight'].values +
              u_week[id_idx] * train_long['week'].values +
              u_sex[id_idx] * train_long['sex'].values +
              u_diet[id_idx] * train_long['diet'].values)
        
        # Likelihood
        rbg_obs = pm.Normal('rbg_obs', mu=mu, sigma=sigma, observed=train_long['rbg'].values)
        trace = pm.sample(1000, tune=1000, cores=1, return_inferencedata=False)

    # Fit Cox Proportional Hazards model for survival data
    cph = CoxPHFitter()
    cph.fit(train_surv, duration_col='Event_time', event_col='Event', formula="sex + diet")
    return trace, cph

# Cross-validation setup
kf = KFold(n_splits=5, shuffle=True, random_state=42)
results = []
for i, (train_index, test_index) in enumerate(kf.split(data['NR_Name'].unique())):
    print(f"Processing fold {i+1}")
    train_ids = data['NR_Name'].unique()[train_index]
    test_ids = data['NR_Name'].unique()[test_index]

    train_long = data[data['NR_Name'].isin(train_ids)]
    test_long = data[data['NR_Name'].isin(test_ids)]
    train_surv = train_long[['NR_Name', 'Event_time', 'Event', 'sex', 'diet']].drop_duplicates()
    test_surv = test_long[['NR_Name', 'Event_time', 'Event', 'sex', 'diet']].drop_duplicates()
    train_overall = train_long.groupby('NR_Name').agg({'rbg': 'mean', 'weight': 'mean', 'week': 'mean'}).reset_index()
    test_overall = test_long.groupby('NR_Name').agg({'rbg': 'mean', 'weight': 'mean', 'week': 'mean'}).reset_index()
    train_overall = pd.merge(train_overall, train_surv, on='NR_Name')
    test_overall = pd.merge(test_overall, test_surv, on='NR_Name')
    trace, cox_fit = joint_model(train_long, train_surv, test_long, test_surv)

    # Predictions for longitudinal data
    test_long = test_long.copy()
    test_long['pred'] = (np.mean(trace['beta_weight']) * test_long['weight'] +
                         np.mean(trace['beta_week']) * test_long['week'] +
                         np.mean(trace['beta_sex']) * test_long['sex'] +
                         np.mean(trace['beta_diet']) * test_long['diet'])
    preds_long = test_long.groupby('NR_Name').agg({'pred': 'mean'}).reset_index()
    preds_long = pd.merge(preds_long, test_overall[['NR_Name', 'rbg']], on='NR_Name')
    rmse = np.sqrt(mean_squared_error(preds_long['rbg'], preds_long['pred']))
    mae = mean_absolute_error(preds_long['rbg'], preds_long['pred'])

    # C-index calculation using lifelines
    preds_surv = test_surv.copy()
    preds_surv['pred'] = cox_fit.predict_partial_hazard(test_surv)
    cindex = cox_fit.concordance_index_
    print(f"Fold {i+1} - RMSE: {rmse}, MAE: {mae}, C-index: {cindex}")
    results.append({'Fold': i+1, 'RMSE': rmse, 'MAE': mae, 'Cindex': cindex})

results_df = pd.DataFrame(results)
print(results_df)
# Note: The above code is a simplified version and may require further adjustments based on the 
# specific dataset and requirements.

File: jointModel.R
#installs and load the package "pacman"
if (!require("pacman")) install.packages("pacman")
pacman::p_load(dplyr, tidyr, ggplot2, nlme,
               survival, JMbayes2, caret, survAUC, pec, timeROC)
library(dplyr)
library(tidyr)
library(ggplot2)
library(nlme)
library(survival)
library(JMbayes2)
library(caret)
library(survAUC)
library(pec)
library(timeROC)

# Read onset_longitudinal
onset_data <- read.csv("onset_longitudinal.csv")
onset <- read.csv("onset.csv")

# Show the number of measurements and unique rats
num_measurements <- nrow(onset_data)
num_unique_rats <- n_distinct(onset_data$NR_Name)
cat("Number of measurements:", num_measurements, "\n")
cat("Number of unique rats:", num_unique_rats, "\n")

# Filter out rows with NA values
onset_data <- drop_na(onset_data)

# Remove rats whose overall diet is Rab from onset_data
rab_overall_rats <- onset %>%
  filter(overall_diet == "Rod" & nursing_diet %in% c("Rab", "Rod")) %>%
  pull(NR_Name)

# Remove rats
onset_data <- onset_data %>% filter(onset_data$NR_Name %in% rab_overall_rats)

# Make a new table grouped by NR_Name and grabbing count
onset_data_grouped <- onset_data %>%
  group_by(NR_Name) %>%
  summarize(num_measurements = n())

# Grab list of rats with only 1 measurement
single_measurement_rats <-
  unique(onset_data_grouped$NR_Name[onset_data_grouped$num_measurements == 1])
# Remove rats with only 1 measurement
onset_data_filtered <- onset_data %>%
  filter(!onset_data$NR_Name %in% single_measurement_rats)

# Remove rows where diet is not Rod or Rab
onset_data_filtered <- onset_data_filtered %>%
  filter(diet %in% c("Rod", "Rab"))

# Show the number of measurements and unique rats
num_measurements <- nrow(onset_data_filtered)
num_unique_rats <- n_distinct(onset_data_filtered$NR_Name)
cat("Number of measurements:", num_measurements, "\n")
cat("Number of unique rats:", num_unique_rats, "\n")

# Make 2 separate tables, one containing the survival data
# And another containing the longitudinal data
# Remove duplicates for survival data
survival_data <- onset_data_filtered %>%
  group_by(NR_Name) %>%
  slice_head(n = 1) %>%
  ungroup() %>%
  select(NR_Name, Event_time, Event, sex) %>%
  merge(onset %>% select(NR_Name, nursing_diet)) %>% # nolint
  rename(diet = "nursing_diet")
longitudinal_data <- onset_data_filtered %>%
  select(NR_Name, week, weight, rbg, sex) %>%
  merge(onset %>% select(NR_Name, nursing_diet)) %>%
  rename(diet = "nursing_diet")

print(table(survival_data$diet))

# Turn chars into factors
longitudinal_data$sex  <- factor(longitudinal_data$sex)
longitudinal_data$diet <- factor(longitudinal_data$diet)

survival_data$sex  <- factor(survival_data$sex,
                             levels = levels(longitudinal_data$sex))
survival_data$diet <- factor(survival_data$diet,
                             levels = levels(longitudinal_data$diet))

# Create 5 folds based on unique NR_Name
set.seed(123)
folds <- createFolds(unique(longitudinal_data$NR_Name), k = 5)

results <- list()
auc <- list()
brier_score <- list()


for (i in seq_along(folds)) {
  print(paste("Processing fold", i))
  test_ids <- unique(longitudinal_data$NR_Name)[folds[[i]]]
  train_ids <- setdiff(unique(longitudinal_data$NR_Name), test_ids)

  train_long <- subset(longitudinal_data, NR_Name %in% train_ids)
  test_long  <- subset(longitudinal_data, NR_Name %in% test_ids)

  train_surv <- subset(survival_data, NR_Name %in% train_ids)
  test_surv  <- subset(survival_data, NR_Name %in% test_ids)

  # Remove diet from test_surv
  test_surv2 <- test_surv %>% select(-diet)
  # Merge test_long and test_surv2 by NR_Name and sex (left join)
  test_overall <- left_join(test_long, test_surv2, by = c("NR_Name", "sex"))
  # Sort test_overall by NR_Name and week
  test_overall <- test_overall %>% arrange(NR_Name, week)

  # Longitudinal mixed effects model
  print(paste("Fitting models for fold", i))
  lme_fit <- lme(rbg ~ weight + week + sex + diet,
                 random = ~ weight + week | NR_Name,
                 data = train_long, control = lmeControl(opt = "optim"))

  cox_fit <- coxph(Surv(Event_time, Event) ~ sex + diet,
                   data = train_surv, x = TRUE)

  print(paste("Fitting joint model for fold", i))
  jm_fit <- jm(cox_fit, lme_fit, time_var = "week")
  print(summary(jm_fit))
  print(coef(jm_fit))

  # Example: prediction on test set
  print(paste("Predicting for fold", i))
  preds_long <- predict(jm_fit, newdata = test_overall,
                        process = "longitudinal")
  preds_surv <- predict(jm_fit, newdata = test_overall, process = "event")

  # Collect results
  rmse <- sqrt(mean((unlist(test_overall$rbg) -
                       unlist(preds_long$pred))^2, na.rm = TRUE))
  mae  <- mean(abs(unlist(test_overall$rbg) -
                     unlist(preds_long$pred)), na.rm = TRUE)

  print("Calculating Brier Score and AUC")

  times <- sort(unique(test_surv$Event_time))

  # Brier Score via pec
  pec_res <- pec::pec(
    object = list("Cox" = cox_fit),
    formula = Surv(Event_time, Event) ~ sex + diet,
    data = test_surv,
    traindata = train_surv,  # note: 'traindata' not 'train_data'
    times = times
  )
  print(pec::ibs(pec_res))  # Integrated Brier Score (one summary number)

  # Time-dependent AUC via timeROC

  lp <- predict(cox_fit, newdata = test_surv, type = "lp")  # linear predictor

  # Pick a random NR_Name from test_surv
  random_rat <- sample(test_surv$NR_Name, 1)

  auc_res <- timeROC::timeROC(
    T = test_surv$Event_time,
    delta = test_surv$Event,
    marker = lp,
    cause = 1,
    times = times,
    iid = FALSE
  )
  print(auc_res$AUC)  # AUC at each time point
  append(auc, auc_res$AUC)

  print("Calculating C-index")
  cindex_res <- pec::cindex(
    object = list("Cox" = cox_fit),
    formula = Surv(Event_time, Event) ~ sex + diet,
    data = test_surv,
    eval.times = sort(unique(test_surv$Event_time)),
    splitMethod = "none"   # manually looping folds
  )

  print(cindex_res$AppCindex)
  fold_results <- c(i, rmse, mae)
  results[[i]] <- fold_results
}

# Combine all fold results into a single data frame
final_results <- do.call(rbind, results)
print(auc)
print("Cross-Validation Results:")
print(final_results)